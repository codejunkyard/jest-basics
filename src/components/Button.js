import React from "react";
import PropTypes from "prop-types";

export default class Button extends React.PureComponent {
  static propTypes = {
    onClick: PropTypes.func,
    disabled: PropTypes.bool
  };

  onClick = event => {
    event.preventDefault();

    if (!this.props.disabled && this.props.onClick) {
      this.props.onClick();
    }
  };

  render() {
    let { children, style = {}, disabled = false } = this.props;

    return (
      <button disabled={disabled} style={style} onClick={this.onClick}>
        {children}
      </button>
    );
  }
}
