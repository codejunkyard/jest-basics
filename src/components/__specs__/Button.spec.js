import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";

import Button from "../Button";

describe("Buttton", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<Button onClick={null}>My button</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render children", () => {
    const wrapper = shallow(<Button>Hello World!</Button>);

    expect(wrapper.text()).toBe("Hello World!");
  });

  it("should render nested children", () => {
    const wrapper = shallow(
      <Button>
        <a href="#">Hello World!</a>
      </Button>
    );

    expect(wrapper.html()).toEqual(
      '<button><a href="#">Hello World!</a></button>'
    );
  });

  describe("onClick", () => {
    const event = { preventDefault: () => {} };

    it("calls props.onClick if it exists", () => {
      const onClick = jest.fn();
      const wrapper = mount(<Button onClick={onClick}>Testing Click</Button>);

      wrapper
        .find("button")
        .hostNodes()
        .simulate("click");
      expect(onClick).toHaveBeenCalled();
    });

    it("calls props.onClick if disabled is false", () => {
      const onClick = jest.fn();
      const wrapper = mount(
        <Button disabled={false} onClick={onClick}>
          Testing Click
        </Button>
      );

      wrapper
        .find("button")
        .hostNodes()
        .simulate("click");
      expect(onClick).toHaveBeenCalled();
    });

    it("should call event.preventDefault if disabled is true", () => {
      const wrapper = shallow(<Button disabled onClick={() => {}} />);
      wrapper.find("button").simulate("click", { preventDefault() {} });
    });

    it("does not call props.onClick if disabled is true", () => {
      const onClick = jest.fn();
      const wrapper = mount(
        <Button disabled onClick={onClick}>
          Testing Click
        </Button>
      );

      wrapper
        .find("button")
        .hostNodes()
        .simulate("click");
      expect(onClick).not.toHaveBeenCalled();
    });
  });
});
