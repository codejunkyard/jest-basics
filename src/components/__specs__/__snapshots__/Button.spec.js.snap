// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttton renders correctly 1`] = `
<button
  disabled={false}
  onClick={[Function]}
  style={Object {}}
>
  My button
</button>
`;
